### PR TITLE
Generator option for mapping proto files to Swift modules

### LINKED ISSUE #294: Story for generating code being used in Swift modules
The current code generation assumes the runtime will come from a module named `SwiftProtobuf` and that all generated code will end up in a single module.  But reality likely won't match that.

Something gRPC likely will have some protos it will bundle, and then the code they generate for a service will likely get built into another module.  So those _service_ will likely need something like _import gRPC_ to get access to the protos shipped there.  And to avoid collisions, any proto that is in _gRPC_ should likely get references in the generated code as _gRPC.[PROTO_NAME]_ (just like we do for the _SwiftProtobuf_ references.

This can also quickly cascade, one package might expose one set of protos, another package might need to reference those, a third provides some independent protos, and then a forth comes in and has protos that reference the ones from two and three.

Initially it might make sense to think of trying to providing a mapping that uses [proto packages](https://developers.google.com/protocol-buffers/docs/proto#packages) to decide what Swift package the code will be in; but that constraint is likely to cause problems.  A collection of proto files might all have the same proto package, but they could be needed by different independent modules, so they want to be in different Swift modules.

For reference the ObjC generator used a [generator options for this](https://github.com/google/protobuf/tree/master/objectivec#objective-c-generator-protoc-options).  Swift being different, the `generate_for_named_framework` doesn't seem like it will apply, as the module the code will end up in hopefully will never be needed because it won't be imported and the types shouldn't need scoping.  So something like `named_framework_to_proto_path_mappings_path` may make the most sense, some form of input let the generator figure out what modules would needed to be imported by each file for the referenced types and how to prefix the types to avoid collisions.

### PR DESCRIPTION
**Pull request by thomasvl**
Fixes #294 

I'm especially happy with the last two commits since it shows how easily this fit into the overhaul of the generator code and the shows it working nicely on all the testing files.

### CODE REVIEW

File: Documentation/PLUGIN.md [hunk 1]
```diff
@@ -123,6 +123,40 @@ The possible values for `Visibility` are:
...
+This options allow you to provide a mapping of module names the and proto files that
+will end up in each one. This data is used during generation to then `import` the
+module and scope the types. This option takes an path to where to read a TextFormat
+mapping:
+
+```
+$ protoc --swift_opt=ProtoPathModuleMappings=[path.asciipb] --swift_out=. foo/bar/*.proto
+```
+
+The format of that mapping files defined in
```
Comment (allevato) [inline Documentation/PLUGIN.md]: s/files/file is/

File: Sources/PluginLibrary/SwiftProtobufNamer.swift [hunk 2]
```diff
@@ -16,9 +16,31 @@ import Foundation
...
+  public convenience init() {
+    self.init(protoFileToModuleMappings: ProtoFileToModuleMappings(), targetModule: "")
+  }
+
+  /// Initializes a a new namer.  All names will be generated as from the pov of the
+  /// given file using the provided file to module mapper.
+  public convenience init(
+    currentFile file: FileDescriptor,
+    protoFileToModuleMappings mappings: ProtoFileToModuleMappings
+   ) {
```
Comment (allevato) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: Extra leading space

File: Tests/PluginLibraryTests/Test_ProtoFileToModuleMappings.swift [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+import XCTest
+@testable import PluginLibrary
+
+// Support equality to simplify testing of getting the correct errors.
+extension ProtoFileToModuleMappings.LoadError: Equatable {
+  public static func ==(lhs: ProtoFileToModuleMappings.LoadError, rhs: ProtoFileToModuleMappings.LoadError) -> Bool {
+    switch (lhs, rhs) {
+    case (.entryMissingModuleName(let l), .entryMissingModuleName(let r)): return l == r
+    case (.entryHasNoProtoPaths(let l), .entryHasNoProtoPaths(let r)): return l == r
+      case (.duplicateProtoPathMapping(let l1, let l2, let l3),
```
Comment (allevato) [inline Tests/PluginLibraryTests/Test_ProtoFileToModuleMappings.swift]: Extra leading space

File: Documentation/PLUGIN.md (hunk 1 above)
Comment (thomasvl) [inline Documentation/PLUGIN.md]: Done

File: Sources/PluginLibrary/SwiftProtobufNamer.swift (hunk 2 above)
Comment (thomasvl) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: Done

File: Tests/PluginLibraryTests/Test_ProtoFileToModuleMappings.swift (hunk 3 above)
Comment (thomasvl) [inline Tests/PluginLibraryTests/Test_ProtoFileToModuleMappings.swift]: Done.

File: Documentation/PLUGIN.md [hunk 4]
```diff
@@ -123,6 +123,40 @@ The possible values for `Visibility` are:
 * `Public`: The visibility on the types is set to `public` so the types will
   be exposed outside the module they are compiled into.
 
+
+##### Generation Option: `ProtoPathModuleMappings` - Swift Module names for proto paths
+
+By default, the sources generated assume all the generated types will be in the same
+module. However, since protos can reference types from another proto file, there is
+the possibility that those generated files might need to end up different modules.
```
Comment (tbkka) [inline Documentation/PLUGIN.md]: "... from another proto file, those generated files might end up in different modules."

File: Documentation/PLUGIN.md [hunk 5]
```diff
@@ -123,6 +123,40 @@ The possible values for `Visibility` are:
...
+This options allow you to provide a mapping of module names the and proto files that
```
Comment (tbkka) [inline Documentation/PLUGIN.md]: This option allows you to specify that the code generated from the proto files will be distributed in multiple modules.

File: Documentation/PLUGIN.md [hunk 6]
```diff
@@ -123,6 +123,40 @@ The possible values for `Visibility` are:
...
+will end up in each one. This data is used during generation to then `import` the
+module and scope the types. This option takes an path to where to read a TextFormat
```
Comment (tbkka) [inline Documentation/PLUGIN.md]: "... takes the path of a file providing the mapping:"

File: Documentation/PLUGIN.md [hunk 7]
```diff
@@ -123,6 +123,40 @@ The possible values for `Visibility` are:
...
+}
+mapping {
+  module_name: "OtherModule"
+  proto_file_path: "mumble.proto"
+  proto_file_path: "other/file.proto"
+}
+```
+
+The `proto_file_path` values should be just like the `import` paths in the proto files,
+they are the how `protoc` sees the file during generation.
```
Comment (tbkka) [inline Documentation/PLUGIN.md]: "The `proto_file_path` values here should match the paths used in the proto file `import` statements."

File: Documentation/PLUGIN.md [hunk 8]
```diff
@@ -123,6 +123,40 @@ The possible values for `Visibility` are:
 * `Public`: The visibility on the types is set to `public` so the types will
   be exposed outside the module they are compiled into.
 
+
+##### Generation Option: `ProtoPathModuleMappings` - Swift Module names for proto paths
+
+By default, the sources generated assume all the generated types will be in the same
```
Comment (tbkka) [inline Documentation/PLUGIN.md]: "By default, the code generator assumes all of the resulting Swift files will be put into the same module."

File: Reference/google/protobuf/test_messages_proto3.pb.swift [hunk 9]
```diff
@@ -499,212 +499,212 @@ struct ProtobufTestMessages_Proto3_TestAllTypes: SwiftProtobuf.Message {
   }
 
   /// Well-known types
-  var optionalBoolWrapper: Google_Protobuf_BoolValue {
-    get {return _storage._optionalBoolWrapper ?? Google_Protobuf_BoolValue()}
+  var optionalBoolWrapper: SwiftProtobuf.Google_Protobuf_BoolValue {
+    get {return _storage._optionalBoolWrapper ?? SwiftProtobuf.Google_Protobuf_BoolValue()}
```
Comment (tbkka) [inline Reference/google/protobuf/test_messages_proto3.pb.swift]: Nice!

File: Documentation/PLUGIN.md (hunk 8 above)
Comment (thomasvl) [inline Documentation/PLUGIN.md]: Done

File: Documentation/PLUGIN.md (hunk 4 above)
Comment (thomasvl) [inline Documentation/PLUGIN.md]: Done

File: Documentation/PLUGIN.md (hunk 5 above)
Comment (thomasvl) [inline Documentation/PLUGIN.md]: Done

File: Documentation/PLUGIN.md (hunk 6 above)
Comment (thomasvl) [inline Documentation/PLUGIN.md]: Done.

File: Documentation/PLUGIN.md (hunk 7 above)
Comment (thomasvl) [inline Documentation/PLUGIN.md]: Done

**Pull request merged**